src/functions/filecheck.sh

```diff
@@ -138,10 +138,18 @@ filecheck() {
   FS_cnt_unchecked=$(grep -cFxf <(sort <<< "${FS_func_libc}") <(sort <<< "${FS_func}"))
   FS_cnt_total=$((FS_cnt_unchecked + FS_cnt_checked))
 
-  if [[ $FS_cnt_checked -eq $FS_cnt_total ]]; then
-    echo_message '\033[32mYes\033[m' 'Yes,' ' fortify_source="yes" ' '"fortify_source":"yes",'
+  if [[ "${FS_cnt_total}" == "0" ]]; then
+    echo_message "\033[32mN/A\033[m" "N/A," ' fortify_source="n/a" ' '"fortify_source":"n/a",'
   else
-    echo_message "\033[31mNo\033[m" "No," ' fortify_source="no" ' '"fortify_source":"no",'
+    if [[ $FS_cnt_checked -eq $FS_cnt_total ]]; then
+      echo_message '\033[32mYes\033[m' 'Yes,' ' fortify_source="yes" ' '"fortify_source":"yes",'
+    else
+      if [[ "${FS_cnt_checked}" == "0" ]]; then
+        echo_message "\033[31mNo\033[m" "No," ' fortify_source="no" ' '"fortify_source":"no",'
+      else
+        echo_message "\033[33mPartial\033[m" "Partial," ' fortify_source="partial" ' '"fortify_source":"partial",'
```

Comment:
Not sure I understand this "Partial" use case?  I understand the N/A but not the partial.  Can you please provide an example or more info on the partial result and a test case for partial?

Reply:
https://github.com/slimm609/checksec.sh/issues/235#issuecomment-2068187001

> Case 2: Fortified greater than zero but less than Fortifiable so partial (analogous to Partial/Full RELRO) -> FORTIFY=Partial

`checksec --file=/usr/bin/ssh`

```
RELRO           STACK CANARY      NX            PIE             RPATH      RUNPATH      Symbols         FORTIFY Fortified       Fortifiable     FILE
Full RELRO      Canary found      NX enabled    PIE enabled     No RPATH   No RUNPATH   No Symbols      Partial 10              21              /usr/bin/ssh
```

Reply:
Do you have any example binaries we could add to the test coverage to validate it?

Reply:
For "Partial" all binaries from my Slackware Linux distribution compiled with "-D_FORTIFY_SOURCE=2"
For the other simple binary cases compiled from [here](https://opensource.com/article/21/6/linux-checksec) (hello.c and hello) and [here](https://medium.com/@slimm609/checksec-d4131dff0fca) (hello1,c, hello1, helllo1f).
The source files and executables used in the testing of the issues #235
[sample.zip](https://github.com/slimm609/checksec.sh/files/15062383/sample.zip)

Reply:
`make test`
doesn't work in Slackware (too many dependencies) nor in Ubuntu 22.04 (I abandoned this yesterday but I could resume it if necessary)

Reply:
I will take a look at the tests and expand.  Please run `make build` and check in that file as well.

Reply:
Slackware64-current and Ubuntu 22.04.4 LTS
`make build`

> Building checksec
> ./hack/build.sh

The modified code in **src/functions/filecheck.sh** is now in **checksec** (generated when running the command)

Reply:
Check for #230 also failed.

Reply:
> `make test`
doesn't work in Slackware (too many dependencies) nor in Ubuntu 22.04 (I abandoned this yesterday but I could resume it if necessary)

Status for: **tests/test-checksec.sh**
xml-checks.sh -> OK
json-checks.sh -> Failed
hardening-checks.sh -> OK (if I remove the cases for "N/A" and "Partial" added by me which are broken)

Reply:
> I will take a look at the tests and expand. Please run `make build` and check in that file as well.

I changed the generation of the **partial, partial_cl, partial32 and partial_cl32** binaries in the **tests/binaries/build_binaries.sh** file by replacing **-D_FORTIFY_SOURCE=2** with **-D_FORTIFY_SOURCE=1** and solved the "**Partial**" case.
I modified accordingly in **tests/hardening-checks.sh**
```
# Partial
for bin in partial partial32 partial_cl partial_cl32; do
  if [[ $("${PARENT}"/checksec --file="${DIR}/binaries/${bin}" --format=csv | cut -d, -f8) != "Partial" ]]; then
    echo "No Fortify validation failed on \"${bin}\": $("${PARENT}"/checksec --file="${DIR}/binaries/${bin}" --format=csv | cut -d, -f8)"
    exit 1
  fi
done
```
I still have no other solution for the "**N/A**" case than adding into the folder **tests/binaries** of the **hello.c** source file from the **sample.zip** archive and generating the binaries using the same compilation options as in the case of **none, none_cl, none32 and none_cl32** binaries followed by the similar modification of the corresponding **for** cycle in **tests/hardening-checks.sh**

Reply:
the rendered checksec file needs checked in.

Reply:
The N/A check needs to be updated a bit more. a binary that has no dependencies on libc should report N/A but the code doesn't currently check against the binary to see if it actually uses libc first. 

The end result is still the same for N/A but extra work that can likely be avoided.   There is a large rewrite in progress for checksec at the moment but it's not a small task so we can fix this for the current version and these improvements are being incorporated into the refactor.

Reply:
> The N/A check needs to be updated a bit more. a binary that has no dependencies on libc should report N/A but the code doesn't currently check against the binary to see if it actually uses libc first.
> The end result is still the same for N/A but extra work that can likely be avoided.

I need help here.

Reply:
I will look at making the updates once this is passing.  I found some other bugs last night testing as well.

Reply:
The solution chosen to test the libc dependency:
`ldd "${1}" 2> /dev/null | grep 'libc\.so' | cut -d' ' -f3`
differs a bit from the one I thought of, that is was:
`${readelf} -d "${1}" 2> /dev/null | grep 'NEEDED' | grep 'libc\.so'`
or (maybe)
`${readelf} -d "${1}" 2> /dev/null | grep 'NEEDED' | grep 'libc\.so' | awk '{print $5}`

Source:
https://www.baeldung.com/linux/show-shared-libraries-executables

> **3. Using the ldd Command**
> ...
> The ldd command is pretty handy to list the shared libraries of a program.
> However, **we should use it with caution, as the ldd utility may execute the program to get the list of the shared libraries. We should never run the ldd command on untrusted executables.**

> **7. Conclusion**
> In this article, we’ve discussed different ways to list shared libraries of a program.
> The ldd command is the most straightforward one to show the shared libraries of a program. The readelf command is a better choice compared to the rest since we deal with “ELF” format on Linux. However, we must keep in mind that we should never use it on untrusted executables.
>